lib/ilp.js

```diff
@@ -15,39 +15,45 @@ class Ilp {
         totalSent: 0
       }
 
-      this.log.info('Creating stream...')
-      const stream = connection.createStream()
-      this.log.info('Stream created successfully')
-
-      stream.on('outgoing_money', (sentAmount) => {
-        const sentAmountInt = parseInt(sentAmount, 10)
-
-        this.log.info('Money sent successfully (%d)!', sentAmountInt)
-
-        state.totalSent += sentAmountInt
-
-        if (state.totalSent === ilpAmount) {
-          state.moneySent = true
-
-          this.log.info('All money is sent. Closing stream...')
-          stream.destroy()
-        } else {
-          this.log.info('%d !== %d... still waiting for all money to send', state.totalSent, ilpAmount)
-        }
-      })
-
-      stream.once('close', () => {
-        this.log.info('Payment stream closed')
-        if (state.moneySent) resolve()
-        else reject(new Error('Stream closed without sending all money'))
-      })
-
-      stream.once('error', (err) => {
-        reject(err)
-      })
-
-      this.log.info('Setting stream max to specified amount (%d)', ilpAmount)
-      stream.setSendMax(ilpAmount)
+      // Wrap this whole thing in a try catch in case it errors and we've partially
+      // sent payments
+      try {
+        this.log.info('Creating stream...')
+        const stream = connection.createStream()
+        this.log.info('Stream created successfully')
+
+        stream.on('outgoing_money', (sentAmount) => {
+          const sentAmountInt = parseInt(sentAmount, 10)
+
+          this.log.info('Money sent successfully (%d)!', sentAmountInt)
+
+          state.totalSent += sentAmountInt
+
+          if (state.totalSent === ilpAmount) {
+            state.moneySent = true
+
+            this.log.info('All money is sent. Closing stream...')
+            stream.destroy()
+          } else {
+            this.log.info('%d !== %d... still waiting for all money to send', state.totalSent, ilpAmount)
+          }
+        })
+
+        stream.once('close', () => {
+          this.log.info('Payment stream closed')
+          if (state.moneySent) resolve({ success: true, remainingAmount: 0 })
+          else resolve({ success: false, remainingAmount: ilpAmount - state.totalSent })
+        })
+
+        stream.once('error', (err) => {
+          resolve({ success: false, remainingAmount: ilpAmount - state.totalSent, error: err.message })
+        })
+
+        this.log.info('Setting stream max to specified amount (%d)', ilpAmount)
+        stream.setSendMax(ilpAmount)
+      } catch (e) {
+        resolve({ success: false, remainingAmount: ilpAmount - state.totalSent, error: e.message })
```

Comment:
in each place where we return the error, return the whole error (not just the message); that way if we rethrow it, we can get the stack trace